Read omdb-movies using resttemplate and DTO

Garret-vin opened the pull request:
Reading movies from rest-api and saving its to database using spring batch

Garret-vin committed:
project without tests
omdb api
spring batch
swagger

Garret-vin committed:
change method name in MovieService
isNotPresent -> existsInDb

Garret-vin committed:
add Language entity
add get from db by language controller

Garret-vin committed:
add api to get by votes
add api to get by rating
small fix in entities

Garret-vin committed:
add JUnit tests to MovieUtil class

Garret-vin committed:
add omdbClient mock tests
add part of mock tests MovieService

Garret-vin committed:
change travis settings

Garret-vin committed:
change travis settings

Garret-vin committed:
add custom exceptions
add tests to movieServiceImpl
refactoring classes + pom

Garret-vin committed:
changed ignore settings for codecov

Garret-vin committed:
was changed:
ignore settings for codecov
settings for jacoco plugin
MovieDatabaseController optimized
add async to spring batch
other small fixes

Garret-vin committed:
add jooq dependency+config plugin
add @Async to omdbController

Garret-vin committed:
add dto entities
add new impl for omdbClient to get all movies from api

Garret-vin committed:
add custom batch reader, processor, writer
delete previous omdb-client impl using omdb-api
add omdb-client impl using resttemplate
add DTO classes for json parsing
other minor refactoring

Garret-vin committed:
add unit tests for ShortMovieReader and ShortMovieToMovieProcessor

mmarusich commented on dao/src/main/java/com/garret/movies/dao/entity/Movie.java:
```diff
@@ -0,0 +1,63 @@
+    @OneToMany(cascade = CascadeType.ALL)
+    private List<Country> countries;
+
+    public Movie(@JsonProperty("Released") String released,
```
@Garret-vin Why DB model has Json creator config?

Garret-vin replied:
> @Garret-vin Why DB model has Json creator config?

@mmarusich Because types of fields in the model are different from json-response. In creator i convert its via MovieUtil class.

mmarusich commented on dao/src/main/java/com/garret/movies/dao/entity/MoviesDto.java:
```diff
@@ -0,0 +1,16 @@
+package com.garret.movies.dao.entity;
```
@Garret-vin not correct package for DTO as for me

mmarusich commented on dao/src/main/java/com/garret/movies/dao/repository/MovieRepository.java:
```diff
@@ -0,0 +1,31 @@
+
+    boolean existsByImdbId(String imdbId);
+
+    int removeById(Long id);
```
such metdhod already exists in CRUD

mmarusich commented on dao/src/main/java/com/garret/movies/dao/repository/MovieRepository.java:
```diff
@@ -0,0 +1,31 @@
+
+    List<Movie> findAllByCountriesName(String country);
+
+    boolean existsByImdbId(String imdbId);
```
same, such method already exists

Garret-vin replied:
@mmarusich no, this method checks by IMDB id, not by id, it's different ids

mmarusich commented on dao/src/main/java/com/garret/movies/dao/util/MovieUtil.java:
```diff
@@ -0,0 +1,117 @@
+
+    private static final SimpleDateFormat formatter = new SimpleDateFormat("dd MMM yyyy", Locale.ENGLISH);
+    private static final Pattern DATE_PATTERN =
+            Pattern.compile("^(([0-9])|([0-2][0-9])|([3][0-1])) (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) \\d{4}$");
```
@Garret-vin Looks odd, what this for?

Garret-vin replied:
@mmarusich this pattern checks for correct input date for parsing in method parseDate()

mmarusich commented on omdb-client/api/src/main/java/com/garret/movies/omdb/api/OmdbClient.java:
```diff
@@ -0,0 +1,13 @@
+
+    List<ShortMovie> searchMovies(String title);
+
+    Movie searchByImdbId(String imdbId);
```
@Garret-vin And this is not cool :( Always split layers. You have API layer and persistance layer. So don't mix them.
You should not work with  DB models here, create own DTO's to work with API

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientDTOImpl.java:
```diff
@@ -0,0 +1,56 @@
+                .queryParam("page", page);
+        while (hasNextPage) {
+            MoviesDto response = restTemplate.getForObject(urlBuilder.toUriString(), MoviesDto.class);
+            if (response != null && response.getResponse().equals("True")) {
```
Please provide http call url so I can  look to JSON we getting from API

Garret-vin replied:
@mmarusich http://www.omdbapi.com?apikey=dfdcd469&r=json&s=batman&page=1

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/config/OmdbConfig.java:
```diff
@@ -0,0 +1,34 @@
+public class OmdbConfig {
+
+    @Bean
+    public RestTemplate restTemplate() {
```
This is regular RestTemplate and does not contains any special for OMDB and in addition you already have bean for rest template in context provided by Spring Boot AutoConfiguration

Garret-vin replied:
@mmarusich I need http converter which is not in default rest template, that's why i configure it

mmarusich commented on pom.xml:
```diff
@@ -42,12 +72,21 @@
+                <configuration>
+                    <excludes>
+                        <exclude>**/com/garret/movies/MovieApplication.class</exclude>
+                        <exclude>**/com/garret/movies/dao/entity/**/*</exclude>
```
https://www.rainerhahnekamp.com/en/ignoring-lombok-code-in-jacoco/

mmarusich commented on web/src/main/java/com/garret/movies/config/SpringBatchConfig.java:
```diff
@@ -0,0 +1,95 @@
+    @StepScope
+    public ItemReader<ShortMovie> itemReader(@Value("#{jobParameters[title]}") String title) {
+        List<ShortMovie> response = omdbClient.searchMovies(title);
+        return new ShortMovieReader(response);
```
@Garret-vin Nope, you should call omdbClient.searchMovies(title) inside reader

mmarusich commented on web/src/main/java/com/garret/movies/config/SpringBatchConfig.java:
```diff
@@ -0,0 +1,95 @@
+    @Bean
+    @StepScope
+    public ItemProcessor<ShortMovie, Movie> itemProcessor() {
+        return new ShortMovieToMovieProcessor(omdbClient, movieService);
```
You already defined it as bean by adding Component, so why doing again here? Choose single approach

mmarusich commented on web/src/main/java/com/garret/movies/controller/MovieDatabaseController.java:
```diff
@@ -0,0 +1,74 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/database-movies")
```
why database? just movies

mmarusich commented on web/src/main/java/com/garret/movies/controller/MovieOmdbDtoController.java:
```diff
@@ -0,0 +1,47 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/v2")
```
why v2? when v1 was deprecated?

Garret-vin replied:
> why v2? when v1 was deprecated?

@mmarusich v1 was without rest template )

mmarusich commented on web/src/main/java/com/garret/movies/controller/MovieDatabaseController.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<Movie> getById(@PathVariable("id") Long id) {
```
@Garret-vin Again about layers, first of all you can't return DB models via API, it can  cause issues with lazy loading and serialization, also you need to make API layer divided from persistence

mmarusich reviewed (changes requested):
@Dominys Please check my comments and make  session with @Garret-vin about that.

Garret-vin committed:
add dto models to movie-service

Dominys commented on movie-service/src/main/java/com/garret/movies/service/util/MovieUtil.java:
```diff
@@ -0,0 +1,117 @@
+            Pattern.compile("[0-9]{1,13}(\\.[0-9]*)?");
+
+    public static List<ActorDto> actorsToList(String actors) {
+        if (actors.isEmpty()) {
```
@Garret-vin This methods are mostly same. Please extract common logic.

Dominys commented on movie-service/src/main/java/com/garret/movies/service/util/MovieUtil.java:
```diff
@@ -0,0 +1,117 @@
+            log.info("Number of votes not found, default value was set");
+            return 0;
+        }
+        String[] splitArray = votes.split(",");
```
@Garret-vin Replace -> parse.

Garret-vin committed:
add lombok.config for jacoco ignore
add entities for omdb-api
add dtos for movie-service
add modelMapper config in web module
exception and MovieUtil moved to web module
fix tests

Garret-vin committed:
add custom BeanPostProcessor for configure ModelMapper
add split configurers for ModelMapper

Garret-vin committed:
change crud repo to jpa repo that would not to convert iterable to list in MovieService
small optimization in methods of MovieService+tests

Garret-vin committed:
added generalization interface to dtos in MovieService
generify method for construct list from string in MovieUtil class
added tests for it

Garret-vin committed:
fix spring batch to do work via multithreading

Garret-vin committed:
added tests to void methods in MovieServiceImpl and MovieDbWriter

Garret-vin committed:
change fields in ActorDto and CountryDto
added 4 modelMapper configurers for these classes
added JsonProperties to dto classes

Garret-vin committed:
added tests to MovieUtil

Garret-vin committed:
added pagination into getAll method
added dto and response for it

Garret-vin committed:
added tests to mvc layer (MovieDatabaseController)

Garret-vin committed:
added atomic index to spring batch processing in ShortMovieReader class

Garret-vin committed:
small fix in ShortMovieReader class

Dominys commented on web/src/main/java/com/garret/movies/batch/ShortMovieReader.java:
```diff
@@ -34,13 +34,7 @@ void init() {
-
-        return shortMovie;
+    public ShortMovie read() {
+        return index.get() < movieData.size() ? movieData.get(index.getAndIncrement()) : null;
```
@Garret-vin OutOfbounds exception goes here ))

Garret-vin replied:
@Dominys I fixed it in last commit

Dominys commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    @Transactional(readOnly = true)
+    public SimpleMoviesResponse getAll(Pageable pageable) {
+        Page<Movie> fullPage = movieRepository.findAll(pageable);
+        Type simpleMovieListType = new TypeToken<List<SimpleMovieDto>>() {
```
@Garret-vin Move to static

Dominys commented on web/src/main/java/com/garret/movies/batch/MovieDbWriter.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public void write(List<? extends OmdbMovie> list) throws Exception {
+        List<OmdbMovie> omdbMovieList = new ArrayList<>(list);
```
@Garret-vin Why you need this?

Garret-vin replied:
@Dominys I needed it earlier, but now i don't need it, so i fixed it

Dominys commented on web/src/main/java/com/garret/movies/util/MovieUtil.java:
```diff
@@ -0,0 +1,73 @@
+                .map(value -> {
+                    T valuable = null;
+                    try {
+                        valuable = clazz.newInstance();
```
@Garret-vin Don't use reflection!!

Dominys reviewed (commented):
Stop adding logic! Even more - split this into separate PR's

Garret-vin committed:
added using switch instead reflection .newInstance() in MovieUtil
fix spring batch writer
moved to static typeToken in MovieServiceImpl

Garret-vin committed:
added using switch instead reflection .newInstance() in MovieUtil
fix spring batch writer
moved to static typeToken in MovieServiceImpl

Garret-vin committed:
Merge remote-tracking branch 'origin/dto_impl' into dto_impl

# Conflicts:
#	web/src/main/java/com/garret/movies/util/MovieUtil.java

Garret-vin committed:
added more tests to MovieUtilTest

Garret-vin committed:
optimized modelMapperConfig
using supplier instead switch in MovieUtil

Garret-vin committed:
last

Garret-vin committed:
rename package in omdb-client api module
